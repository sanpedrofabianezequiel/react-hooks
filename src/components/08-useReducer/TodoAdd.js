import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [ {description}, handleInputChange , reset] = useForm({
        description:'',
    });

    const handleSubmit = (e)=>{
        e.preventDefault();
        
        if(description.trim().length <= 1 )
            return;    
        
        const newTodo  = {
            id: new Date().getTime(),
            desc: description,
            done:false
        };         

        
        handleAddTodo(newTodo);
        //this function clear the input
        reset();
    }

    return (
        <>
             <h4>Agregar TODO</h4>
                    <hr />

            <form onSubmit={handleSubmit}>
                        <input onChange={handleInputChange} value={description} type="text" name="description" className="form-control" placeholder="Aprender..." autoComplete="false"/>
                        <button className="btn btn-outline-primary mt-1 btn-block w-100" type="submit">
                            Agregar
                        </button>
            </form>   
        </>
    )
}
