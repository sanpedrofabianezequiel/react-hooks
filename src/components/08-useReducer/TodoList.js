import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({state,handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
        {
            state.map( (todo,i)=>
                    ( 
                        <TodoListItem 
                            key={todo.id}
                            todo={todo}
                            i={i}
                            handleDelete={handleDelete}
                            handleToggle={handleToggle}
                        />
                    )
            )
        }
        </ul>
    )
}
