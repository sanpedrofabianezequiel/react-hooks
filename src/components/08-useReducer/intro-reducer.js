//Crear un objeto inicial para el todoReducer
//llamarlo e inicializarlo con LET
//enviarle al todoReducer  un  state + action  => action debe tener  type + payload 
const initialState =[{
    id:1,
    todo:"Comprar pan",
    done:false,
}];


const todoReducer = (state = initialState, action)=>{

    if(action?.type === 'add'){
        return [...state , action.payload];
    }

    return state;
}

const newTodo =[{
    id:2,
    todo:"Comprar leche",
    done:false,
}];

let todoInit = todoReducer();   //Inicializacion

const action ={
    type:'add',
    payload: newTodo,
}

console.log(todoReducer(todoInit ,action));