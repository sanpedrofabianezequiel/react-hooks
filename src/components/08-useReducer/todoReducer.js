
export const todoReducer = (state =[] ,action)=>{
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter(todo => todo.id !== action.payload );


        case 'toggle':
            return state.map((x)=>
                (x.id===action.payload)
                ? { ...x, done: !x.done}
                : x
            );

        case 'toggle-old':
            return state.map((x)=>{
                if(x.id===action.payload){
                    return {
                        ...x,
                        done: !x.done
                    }
                }else{
                    return x;
                }
            });
        default:
            return state;
    }
}