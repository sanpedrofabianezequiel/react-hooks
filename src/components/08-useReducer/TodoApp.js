import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'
import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init  = ()=>{
   
    return JSON.parse(localStorage.getItem('state')) || [];

    /* return[{
        id: new Date().getTime(),
        desc:'Aprender React',
        done:false
    }];*/
} 

    

export const TodoApp = () => {

    //el primer parametro del reducer es el promiamente dicho
    //el reducer con el state y la accion
    const [state, dispatch] = useReducer(todoReducer, [], init);
  



    useEffect(() => {
        localStorage.setItem('state',JSON.stringify(state));    
    }, [state]);


    const handleAddTodo= (newTodo)=>{
        dispatch({
            type:'add',
            payload:newTodo
        });
    }

   

    const handleDelete = (todoId) =>{
        //1.Create Model
        //2.Create Action
        const action = {
            type: 'delete',
            payload:todoId
        }
        dispatch(action);
    }

    const handleToggle=(todoId)=>{
        dispatch({
            type:'toggle',
            payload:todoId
        });
    }
    return (
        <div>
            <h1> ({ state.length })</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                      <TodoList 
                        state={state}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                      />
                </div>
                <div className="col-5">
                    <TodoAdd 
                    
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>
        </div>
    )
}
