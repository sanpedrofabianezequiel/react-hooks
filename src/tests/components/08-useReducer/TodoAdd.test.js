import { shallow } from 'enzyme'
import React from 'react'
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd'

describe('Pruebas en <TodoAdd/> ', () => {
    
    const handleAddTodo = jest.fn();

    const wrapper = shallow(<TodoAdd
        handleAddTodo={handleAddTodo}
    />);
    test('Debe de mostrar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('no debe de llamar handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test('debe de llamar la funcion handleAddTodo', () => {
       
        const value = "Aprender React";
        wrapper.find('input').simulate('change',{
            target:{
                value:value,
                name:'description'
            }
        });//el segundo parama es el target

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));

        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc:value,
            done:false
        });

        expect(wrapper.find('input').props().value ).toBe('');
    });
    
})
