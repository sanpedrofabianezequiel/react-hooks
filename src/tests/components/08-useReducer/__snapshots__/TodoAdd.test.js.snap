// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd/>  Debe de mostrar correctamente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="false"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block w-100"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
